server/interpreter/data/boxes.go

```diff
@@ -0,0 +1,237 @@
+	if key == "length" {
+		return &ErrorMsg{"TypeError",
+			fmt.Sprintf("Cannot assign to read only property '%s' of %s",
+				key, bstr.ToString())}
+	}
+	return bstr.object.Set(key, value)
+}
+
+// Delete on BoxedString will refuse to delete "length" and numeric
+// character indicies, and otherwise defers to the embedded object.
+//
+// BUG(cpcallen): character indicides not implemented.
+func (bstr *BoxedString) Delete(key string) *ErrorMsg {
+	if key == "length" {
+		return &ErrorMsg{"TypeError",
+			fmt.Sprintf("Cannot delete property '%s' of %s",
+				key, bstr.ToString())}
+	}
+	return bstr.object.Delete(key)
+}
+
+// OwnPropertyKeys on BoxedString returns the length and numeric
+// character indicies plus any properties defined on the embedded
+// object.
+//
+// BUG(cpcallen): character indicides not implemented.
+func (bstr *BoxedString) OwnPropertyKeys() []string {
+	return append([]string{"length"}, bstr.object.OwnPropertyKeys()...)
+}
+
+// HasOwnProperty on BoxedString returns true for "length" and false for all
+// other inputs for BoxedStrings.
+//
+// BUG(cpcallen): character indicides not implemented.
+func (bstr *BoxedString) HasOwnProperty(key string) bool {
+	if key == "length" {
+		return true
+	}
+	return bstr.object.HasOwnProperty(key)
+}
+
+// HasProperty on BoxedString returns true if the specified property
+// name exists on the object or its prototype chain.
+//
+// This must be ~redundantly defined on BoxedString, because the
+// inherited definition from the embedded object would call
+// bstr.object.HasOwnProperty() instead of bstr.HasOwnProperty() and
+// thus not see length etc.
+func (bstr *BoxedString) HasProperty(key string) bool {
+	return bstr.HasOwnProperty(key) || bstr.Proto().HasProperty(key)
+}
+
+// ToBoolean has no special behaviour on BoxedStrings: it returns
+// true, as for any other object.
+
+// ToNumber on BoxedString just returns the same value as on the boxed
+// String.
+func (bstr *BoxedString) ToNumber() Number {
+	return bstr.value.ToNumber()
+}
+
+// ToString on BoxedString just returns the boxed String itself.
+//
+// BUG(cpcallen): ToString should call a user-code toString() method
+// if present.
+func (bstr *BoxedString) ToString() String {
+	return bstr.value
+}
+
+// ToPrimitive returns the contained String on BoxedString objects.
+func (bstr *BoxedString) ToPrimitive() Value {
+	return bstr.value
+}
+
+// NewBoxedString creates a new object with the specified owner and
+// prototype, initialises it as appropriate, and returns a pointer to
+// the newly-created object.
+func NewBoxedString(owner *Owner, proto Object, value String) *BoxedString {
+	var bstr = new(BoxedString)
+	bstr.init(owner, proto)
+	bstr.f = true
+	bstr.value = value
+	return bstr
+}
+
+/********************************************************************/
+
+// Coerce coerces its first argument into an object.
+func Coerce(value Value, owner *Owner, protos *Protos) Object {
```

Comment:
This is probably fine, but since this depends on a set of builtins, it might work better to have this be a method of Interpreter (or some other runtime-global object) so you don't have to thread protos through every function.